prism-agent/service/server/src/test/scala/io/iohk/atala/iam/authentication/oidc/KeycloakAuthenticatorSpec.scala

```diff
@@ -0,0 +1,262 @@
+package io.iohk.atala.iam.authentication.oidc
+
+import io.iohk.atala.agent.walletapi.crypto.ApolloSpecHelper
+import io.iohk.atala.agent.walletapi.model.Wallet
+import io.iohk.atala.agent.walletapi.service.WalletManagementService
+import io.iohk.atala.agent.walletapi.service.WalletManagementServiceImpl
+import io.iohk.atala.agent.walletapi.sql.JdbcWalletNonSecretStorage
+import io.iohk.atala.agent.walletapi.sql.JdbcWalletSecretStorage
+import io.iohk.atala.iam.authentication.AuthenticationError
+import io.iohk.atala.shared.models.WalletId
+import io.iohk.atala.sharedtest.containers.KeycloakAdminClient
+import io.iohk.atala.sharedtest.containers.KeycloakContainerCustom
+import io.iohk.atala.sharedtest.containers.KeycloakTestContainerSupport
+import io.iohk.atala.sharedtest.containers.PostgresTestContainerSupport
+import io.iohk.atala.test.container.DBTestUtils
+import org.keycloak.authorization.client.AuthzClient
+import org.keycloak.representations.idm.CredentialRepresentation
+import org.keycloak.representations.idm.UserRepresentation
+import org.keycloak.representations.idm.authorization.ResourceRepresentation
+import org.keycloak.representations.idm.authorization.UmaPermissionRepresentation
+import zio.*
+import zio.http.Client
+import zio.test.*
+import zio.test.Assertion.*
+
+import java.net.URI
+import scala.jdk.CollectionConverters.*
+
+object KeycloakAuthenticatorSpec
+    extends ZIOSpecDefault,
+      KeycloakTestContainerSupport,
+      PostgresTestContainerSupport,
+      ApolloSpecHelper {
+
+  private def keycloakConfigLayer(authUpgradeToRPT: Boolean = true) =
+    ZLayer.fromZIO {
+      ZIO.serviceWith[KeycloakContainerCustom] { container =>
+        val host = container.container.getHost()
+        val port = container.container.getHttpPort()
+        val url = s"http://${host}:${port}"
+        KeycloakConfig(
+          enabled = true,
+          keycloakUrl = URI(url).toURL(),
+          realmName = realmName,
+          clientId = agentClientRepresentation.getClientId(),
+          clientSecret = agentClientSecret,
+          autoUpgradeToRPT = authUpgradeToRPT
+        )
+      }
+    }
+
+  private def createWalletResource(walletId: WalletId, name: String) =
+    for {
+      authzClient <- ZIO.service[AuthzClient]
+      resource = {
+        val resource = ResourceRepresentation()
+        resource.setId(walletId.toUUID.toString())
+        resource.setName(name)
+        resource.setOwnerManagedAccess(true)
+        resource
+      }
+      _ <- ZIO
+        .attemptBlocking(authzClient.protection().resource().create(resource))
```

Comment:
do we need to use the KeycloakAdminClient if the same resource can be created by the regular client?

Reply:
No, the `authz` client expose the UMA capability. Admin client expose proprietary admin endpoints.